File: medium/235-lowest-common-ancestor-of-a-binary-search-tree/235-lowest-common-ancestor-of-a-binary-search-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */

/**
 * @param {TreeNode} root
 * @param {TreeNode} p
 * @param {TreeNode} q
 * @return {TreeNode}
 */
const lowestCommonAncestor = function(root, p, q) {
    let curr = root;
    while (curr) {
        if (curr.val > p.val && curr.val > q.val) {
            curr = curr.left;
        }
        else if (curr.val < p.val && curr.val < q.val) {
            curr = curr.right
        }
        else return curr;
    }
};

File: medium/235-lowest-common-ancestor-of-a-binary-search-tree/235-lowest-common-ancestor-of-a-binary-search-tree.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

# @param {TreeNode} root
# @param {TreeNode} p
# @param {TreeNode} q
# @return {TreeNode}
def lowest_common_ancestor(root, p, q)
    curr = root
    
    while curr do  
        if curr.val < p.val && curr.val < q.val
            curr = curr.right
        elsif curr.val > p.val && curr.val > q.val
            curr = curr.left
        else
            return curr
        end
    end
end